app/models/mixins/supports_feature_mixin.rb

```diff
@@ -141,72 +180,50 @@ def unsupported
     # This is the DSL used a class level to define what is supported
     def supports(feature, &block)
       SupportsFeatureMixin.guard_queryable_feature(feature)
-      define_supports_feature_methods(feature, &block)
+      supported_feature_definitions[feature.to_sym] = FeatureDefinition.new(supported: true, block: block)
     end
 
     # supports_not does not take a block, because its never supported
     # and not conditionally supported
     def supports_not(feature, reason: nil)
       SupportsFeatureMixin.guard_queryable_feature(feature)
-      define_supports_feature_methods(feature, :is_supported => false, :reason => reason)
+      supported_feature_definitions[feature.to_sym] = FeatureDefinition.new(unsupported_reason: reason)
     end
 
     # query the class if the feature is supported or not
     def supports?(feature)
       SupportsFeatureMixin.guard_queryable_feature(feature)
-      public_send("supports_#{feature}?")
+      feature_definition_for(feature).supported?
     end
 
     # query the class for the reason why something is unsupported
     def unsupported_reason(feature)
       SupportsFeatureMixin.guard_queryable_feature(feature)
-      feature = feature.to_sym
-      public_send("supports_#{feature}?") unless unsupported.key?(feature)
-      unsupported[feature]
+      feature_definition_for(feature).unsupported_reason
     end
 
     # query the class if a feature is generally known
     def feature_known?(feature)
       SupportsFeatureMixin::QUERYABLE_FEATURES.key?(feature.to_sym)
     end
 
-    private
-
-    def unsupported
-      # This is a class variable and it might be modified during runtime
-      # because we dont eager load all classes at boot time, so it needs to be thread safe
-      @unsupported ||= Concurrent::Hash.new
-    end
-
-    # use this for making a class not support a feature
-    def unsupported_reason_add(feature, reason = nil)
-      SupportsFeatureMixin.guard_queryable_feature(feature)
+    def feature_definition_for(feature)
       feature = feature.to_sym
-      unsupported[feature] = SupportsFeatureMixin.reason_or_default(reason)
+      feature_definition = nil
+      ancestors.detect do |ancestor|
+        feature_definition = ancestor.try(:supported_feature_definition, feature)
+      end
+      feature_definition || FeatureDefinition.new
     end
 
-    def define_supports_feature_methods(feature, is_supported: true, reason: nil, &block)
-      method_name = "supports_#{feature}?"
-      feature = feature.to_sym
+    def supported_feature_definition(feature)
+      supported_feature_definitions[feature]
+    end
 
-      # defines the method on the instance
-      define_method(method_name) do
-        unsupported.delete(feature)
-        if block_given?
-          instance_eval(&block)
-        else
-          unsupported_reason_add(feature, reason) unless is_supported
-        end
-        !unsupported.key?(feature)
-      end
+    private
```

Comment:
Wonder why rubocop is complaining about this private (perhaps you have a second one somewhere earlier)?

@jrafanie Check it out...code climate engines to the rescue :)